### PR TITLE
[OAuth] sync oauth apps with Central

### PR DESCRIPTION
**Pull request by alrocar**
cartodb part of https://github.com/CartoDB/cartodb/issues/14493

Test together with: https://github.com/CartoDB/cartodb-central/pull/2388

Some decisions taken:

- `user_id` is nullable in `oauth_apps` to allow sync to all the clouds.
- `avoid_sync_central` in `oauth_app` is used to avoid circular synchronizations. It can be also handy for manual operations.
- Sync with central is synchronous. It adds a bit of latency but it's easier to handle and show errors to a user.

### CODE REVIEW

File: app/controllers/superadmin/oauth_apps_controller.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  def oauth_params
+    sync_params = @user ? Carto::OauthApp::ALLOWED_SYNC_PARAMS : Carto::OauthApp::ALLOWED_SYNC_PARAMS - %i(user_id)
+    params[:oauth_app].permit(sync_params)
+                      .merge(avoid_sync_central: true)
+  end
+
+  def avoid_sync_central
+    @oauth_app.avoid_sync_central = true
```
Comment (gonzaloriestra) [inline app/controllers/superadmin/oauth_apps_controller.rb]: You are already setting this as true in `oauth_params`. Isn't this redundant?

File: spec/models/carto/oauth_app_spec.rb [hunk 2]
```diff
@@ -85,6 +87,157 @@ module Carto
...
+        Cartodb::Central.stubs(:sync_data_with_cartodb_central?).returns(false)
+        @oauth_app.destroy! if @oauth_app
+        @oauth_app2.destroy! if @oauth_app2
+      end
+
+      after(:all) do
+        @user_oauth.destroy!
+      end
+
+      describe '#create' do
```
Comment (gonzaloriestra) [inline spec/models/carto/oauth_app_spec.rb]: These specs are weird because they don't have expectations. I guess they are supposed to pass when they don't raise an error, but you could write that with something like: `expect { whatever }.not_to raise_error`.

Or you could check that the element was really saved.

File: app/controllers/superadmin/oauth_apps_controller.rb (hunk 1 above)
Comment (alrocar) [reply app/controllers/superadmin/oauth_apps_controller.rb]: Right, I can remove it on update at least

Comment (gonzaloriestra) [reply app/controllers/superadmin/oauth_apps_controller.rb]: Ah, you need it for the destroy... And what about removing the last merge in `oauth_params` instead?

Comment (alrocar) [reply app/controllers/superadmin/oauth_apps_controller.rb]: I need it for the `create` method 😄

Comment (alrocar) [reply app/controllers/superadmin/oauth_apps_controller.rb]: well actually I can do a `new` and then `save!` 🤔

Comment (gonzaloriestra) [reply app/controllers/superadmin/oauth_apps_controller.rb]: Or you could keep it in `oauth_params` (for create and update), and just change it manually in destroy: 
```
  def destroy
    @oauth_app.avoid_sync_central = true
    @oauth_app.destroy!
    ...
```
As you prefer!

File: app/models/carto/oauth_app.rb [hunk 3]
```diff
@@ -10,17 +10,35 @@ class OauthApp < ActiveRecord::Base
...
 
-    validates :user, presence: true
+    validates :user, presence: true, if: :sync_with_central?
     validates :name, presence: true
     validates :client_id, presence: true
     validates :client_secret, presence: true
     validates :redirect_uris, presence: true
     validates :icon_url, presence: true
-    validates :oauth_app_organizations, absence: true, unless: :restricted?
+    validates :oauth_app_organizations, absence: true, unless: :needs_org?
```
Comment (javitonino) [inline app/models/carto/oauth_app.rb]: Why this change? In any case, a restricted application should not have relation with any organization. For reference:
- Restricted = true -> only the specified organizations can access it. This field may be present (or empty if no one has access in this cloud)
- Restricted = false -> everyone can access it, so this relation should be empty.

File: app/models/carto/oauth_app.rb [hunk 4]
```diff
@@ -10,17 +10,35 @@ class OauthApp < ActiveRecord::Base
...
     validate :validate_uris
 
     before_validation :ensure_keys_generated
 
+    after_save :sync_with_central, if: :sync_with_central?
+    after_destroy :delete_from_central, if: :sync_with_central?
+
+    ALLOWED_SYNC_ATTRIBUTES = %i[id name client_id client_secret redirect_uris icon_url restricted].freeze
+    ALLOWED_SYNC_PARAMS = ALLOWED_SYNC_ATTRIBUTES + [:user_id, redirect_uris: []].freeze
```
Comment (javitonino) [inline app/models/carto/oauth_app.rb]: Input validation is a controller concern, so I think it makes more sense to keep it there. Not totally certain.

File: db/migrate/20181130103140_allow_nulls_in_user_id.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
```
Comment (javitonino) [inline db/migrate/20181130103140_allow_nulls_in_user_id.rb]: Added tag to not merge until the migration has been split.

File: app/models/carto/oauth_app.rb (hunk 3 above)
Comment (alrocar) [reply app/models/carto/oauth_app.rb]: Oh OK, I thought there was a relation.

File: app/models/carto/oauth_app.rb [hunk 6]
```diff
@@ -40,5 +58,38 @@ def validate_uri(redirect_uri)
     rescue URI::InvalidURIError
       errors.add(:redirect_uris, "must be valid")
     end
+
+    def sync_with_central
+      if id_changed?
```
[the remark this answers is not in the document]
Comment (alrocar) [reply app/models/carto/oauth_app.rb]: I was using `id_changed?` because I'm preserving IDs between central and clouds, and once you set the ID for a new record it's not a `new_record?` any more.

[the remark this answers is not in the document]
Comment (javitonino) [reply app/models/carto/oauth_app.rb]: Arghhh. Classic. Then, why not split this into two hooks? `after_create` and `after_update` Maybe that would work?

[the remark this answers is not in the document]
Comment (alrocar) [reply app/models/carto/oauth_app.rb]: Seems to work, thanks!

File: app/models/carto/oauth_app.rb [hunk 7]
```diff
@@ -10,7 +10,7 @@ class OauthApp < ActiveRecord::Base
     has_many :oauth_app_users, inverse_of: :oauth_app, dependent: :destroy
     has_many :oauth_app_organizations, inverse_of: :oauth_app, dependent: :destroy
 
-    validates :user, presence: true
+    validates :user, presence: true, if: :sync_with_central?
```
Comment (javitonino) [inline app/models/carto/oauth_app.rb]: This is incorrect in the case where central is not configured. In that case, this field should always be mandatory.

File: app/models/carto/oauth_app.rb [hunk 8]
```diff
@@ -10,7 +10,7 @@ class OauthApp < ActiveRecord::Base
     has_many :oauth_app_users, inverse_of: :oauth_app, dependent: :destroy
     has_many :oauth_app_organizations, inverse_of: :oauth_app, dependent: :destroy
 
-    validates :user, presence: true, if: :sync_with_central?
+    validates :user, presence: true, if: -> { sync_with_central? || !central_enabled? }
```
Comment (javitonino) [inline app/models/carto/oauth_app.rb]: I'd rewrite this as `unless: -> { central_enabled? && avoid_sync_central }`, but up to you.

**Comment by gonzaloriestra** [comment]
:heavy_check_mark: Acceptance OK

Tested from console:
- CRUD from one cloud to check synchronization in central and other clouds
- CRUD with different flags: `user_id`, `avoid_sync_central`, `restricted`
- CRUD with central disabled
